server/controllers/finance/accounts/references.compute.js

```diff
@@ -0,0 +1,184 @@
+  const glb = {};
+
+  // get fiscal year information for the given period
+  const queryFiscalYear = `
+    SELECT fy.id, p.number AS period_number FROM fiscal_year fy
+    JOIN period p ON p.fiscal_year_id = fy.id
+    WHERE p.id = ?
+  `;
+
+  const queryAccountReference = `
+    SELECT id, abbr, description, is_amo_dep FROM account_reference
+    WHERE abbr = ? AND is_amo_dep = ?;
+  `;
+
+  return db.one(queryFiscalYear, [periodId])
+    .then(fiscalYear => {
+      glb.fiscalYear = fiscalYear;
+
+      return db.one(queryAccountReference, [abbr, isAmoDep]);
+    })
+    .then(ar => {
+      return getValueForReference(
+        ar.abbr,
+        ar.isAmoDep,
+        ar.description,
+        glb.fiscalYear.period_number,
+        glb.fiscalYear.id
+      );
+    })
+    .then(data => {
+      return data[0];
+    });
+}
+
+/**
+ * @method getValueForReference
+ *
+ * @description
+ * Returns computed value of the reference in a given period and fiscal_year
+ *
+ * @param {number} fiscalYearId
+ * @param {number} periodNumber
+ * @param {string} abbr - the reference of accounts. ex. AA or AX
+ * @param {boolean} isAmoDep - the concerned reference is for amortissement, depreciation or provision
+ */
+function getValueForReference(abbr, isAmoDep = 0, referenceDescription, periodNumber, fiscalYearId) {
+  const queryTotals = `
+  SELECT abbr, is_amo_dep, description,
+    IFNULL(debit, 0) AS debit, IFNULL(credit, 0) AS credit, IFNULL(balance, 0) AS balance FROM (
+      SELECT ? AS abbr, ? AS is_amo_dep, ? AS description,
+        SUM(IFNULL(pt.debit, 0)) AS debit, SUM(IFNULL(pt.credit, 0)) AS credit,
+        SUM(IFNULL(pt.debit - pt.credit, 0)) AS balance
+      FROM period_total pt 
+      JOIN period p ON p.id = pt.period_id
+      WHERE pt.fiscal_year_id = ? AND pt.locked = 0 AND p.number BETWEEN 0 AND ? AND pt.account_id IN (?)
+    )z
+  `;
+
+  return getAccountsForReference(abbr, isAmoDep)
+    .then(accounts => {
+      const accountIds = accounts.map(a => a.account_id);
+      const parameters = [
+        abbr,
+        isAmoDep,
+        referenceDescription,
+        fiscalYearId,
+        periodNumber,
+        accountIds.length ? accountIds : null,
+      ];
+      return db.exec(queryTotals, parameters);
+    });
+}
+
+/**
+ * @method getAccountsForReference
+ *
+ * @description
+ * Returns all accounts concerned by the reference without exception accounts
+ *
+ * @param {string} abbr - the reference of accounts. ex. AA or AX
+ * @param {boolean} isAmoDep - the concerned reference is for amortissement, depreciation or provision
+ */
+function getAccountsForReference(abbr, isAmoDep = 0) {
+  /**
+   * Get the list of accounts of the reference without excepted accounts
+   * mysql implementation of minus operator
+   * @link http://www.mysqltutorial.org/mysql-minus/
+   */
+  const queryAccounts = `
+    SELECT includeTable.account_id, includeTable.account_number FROM (
```

Comment:
This is a really complicated query.  Can you walk me through the logic we are trying to get?  It looks like we could simplify it, but I'm not sure what the goal is.